server/message_dispatch.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strings"
+	"sync"
+)
+
+type MessageDispatch struct {
+	input       chan string
+	output      map[string][]chan string
```

Comment:
You're right about buffered channels, I'll apply this change. 

As for using Conds, I tend to prefer high-level abstractions (like chan) over low-level primitives (like Cond+Mutex+buffer). As far as I understand, what you're suggesting is actually encapsulating a Cond (plus helper variables) into a chan. IMO, it'll be a better practice to use ready-made library types instead.

What's more, when designing a prototype with manual synchronization we've came up with a code that was hard to understand and analyze. The problem were sequences of interleaved lock-unlock calls on different objects.